src/librustc_mir/transform/split_local_fields.rs

```diff
@@ -0,0 +1,397 @@
+use crate::transform::{MirPass, MirSource};
+use rustc::mir::visit::{MutVisitor, NonUseContext, PlaceContext, Visitor};
+use rustc::mir::*;
+use rustc::ty::layout::VariantIdx;
+use rustc::ty::util::IntTypeExt;
+use rustc::ty::{self, Ty, TyCtxt};
+use rustc_index::vec::IndexVec;
+use std::collections::BTreeMap;
+use std::iter::Step;
+use std::ops::Range;
+use syntax_pos::Span;
+
+pub struct SplitLocalFields;
+
+impl MirPass<'tcx> for SplitLocalFields {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, _: MirSource<'tcx>, body: &mut BodyAndCache<'tcx>) {
+        let mut collector = FieldTreeCollector {
+            tcx,
+            locals: body.local_decls.iter().map(|decl| FieldTreeNode::new(decl.ty)).collect(),
+        };
+
+        // Can't split return and arguments.
+        collector.locals[RETURN_PLACE].make_opaque();
+        for arg in body.args_iter() {
+            collector.locals[arg].make_opaque();
+        }
+
+        collector.visit_body(read_only!(body));
+
+        let replacements = collector
+            .locals
+            .iter_enumerated_mut()
+            .map(|(local, root)| {
+                // Don't rename locals that are entirely opaque.
+                match root.kind {
+                    FieldTreeNodeKind::Opaque { .. } => local..local.add_one(),
+                    FieldTreeNodeKind::Split(_) => {
+                        let source_info = body.local_decls[local].source_info;
+                        let first = body.local_decls.next_index();
+                        root.assign_locals(&mut body.local_decls, source_info);
+                        first..body.local_decls.next_index()
+                    }
+                }
+            })
+            .collect::<IndexVec<Local, Range<Local>>>();
+
+        // Expand `Storage{Live,Dead}` statements to refer to the replacement locals.
+        for bb in body.basic_blocks_mut() {
+            bb.expand_statements(|stmt| {
+                let (local, is_live) = match stmt.kind {
+                    StatementKind::StorageLive(local) => (local, true),
+                    StatementKind::StorageDead(local) => (local, false),
+                    _ => return None,
+                };
+                let range = replacements[local].clone();
+                // FIXME(eddyb) `Range<Local>` should itself be iterable.
+                let range = (range.start.as_u32()..range.end.as_u32()).map(Local::from_u32);
+                let source_info = stmt.source_info;
+                Some(range.map(move |new_local| Statement {
+                    source_info,
+                    kind: if is_live {
+                        StatementKind::StorageLive(new_local)
+                    } else {
+                        StatementKind::StorageDead(new_local)
+                    },
+                }))
+            });
+        }
+        drop(replacements);
+
+        // Lastly, replace all the opaque nodes with their new locals.
```

Comment:
s/opaque/split

Reply:
Nope, the opaque nodes are the leaves, split nodes don't get a local.

I need to find better terminology for all of this so it's easier to follow, tbh.